Sources/History/HistoryStore.swift

```diff
@@ -0,0 +1,446 @@
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Common
+import Foundation
+import CoreData
+import Combine
+
+final public class HistoryStore: HistoryStoring {
+
+    public enum HistoryStoreEvents {
+
+        case removeFailed
+        case reloadFailed
+        case cleanEntriesFailed
+        case cleanVisitsFailed
+        case saveFailed
+        case insertVisitFailed
+        case removeVisitsFailed
+
+    }
+
+    let context: NSManagedObjectContext
+    let eventMapper: EventMapping<HistoryStoreEvents>
+
+    public init(context: NSManagedObjectContext, eventMapper: EventMapping<HistoryStoreEvents>) {
+        self.context = context
+        self.eventMapper = eventMapper
+    }
+
+    enum HistoryStoreError: Error {
+        case storeDeallocated
+        case savingFailed
+    }
+
+    public func removeEntries(_ entries: [HistoryEntry]) -> Future<Void, Error> {
+        return Future { [weak self] promise in
+            self?.context.perform {
+                guard let self = self else {
+                    promise(.failure(HistoryStoreError.storeDeallocated))
+                    return
+                }
+
+                let identifiers = entries.map { $0.identifier }
+                switch self.remove(identifiers, context: self.context) {
+                case .failure(let error):
+                    self.context.reset()
+                    promise(.failure(error))
+                case .success:
+                    promise(.success(()))
+                }
+            }
+        }
+    }
+
+    public func cleanOld(until date: Date) -> Future<BrowsingHistory, Error> {
+        return Future { [weak self] promise in
+            self?.context.perform {
+                guard let self = self else {
+                    promise(.failure(HistoryStoreError.storeDeallocated))
+                    return
+                }
+
+                switch self.clean(self.context, until: date) {
+                case .failure(let error):
+                    self.context.reset()
+                    promise(.failure(error))
+                case .success:
+                    let reloadResult = self.reload(self.context)
+                    promise(reloadResult)
+                }
+            }
+        }
+    }
+
+    private func remove(_ identifiers: [UUID], context: NSManagedObjectContext) -> Result<Void, Error> {
+        // To avoid long predicate, execute multiple times
+        let chunkedIdentifiers = identifiers.chunked(into: 100)
```

Comment:
Feels like we could use array slices and iterate over windows of the array here rather than allocate a new array – pretty sure UUIDs are value types so I'd expect `chunked` is copying them needlessly.

Reply:
This was copied from macOS and just modified to not be encrypted, sorry I should have made that clearer. I didn't want to change too much and we already discussed the possibility of refactoring this entire part of the stack to make it more efficient anyway.